stock_storage_type/models/stock_location.py

```diff
@@ -475,6 +498,36 @@ def select_allowed_locations(
         )
         return valid_locations
 
+    def _order_allowed_locations(self, valid_locations):
+        """Return the ordered list of valid_locations
+
+        By default the order should be the same as self. However, if the
+        valid_locations list contains locations configured to not mix products,
+        we must give priority to locations that already contains products
+        (the ones with less qty first)
+        """
+        valid_no_mix = valid_locations.filtered("do_not_mix_products")
+        loc_ordered_by_qty = []
+        if valid_no_mix:
+            StockQuant = self.env["stock.quant"]
+            domain_quant = [("location_id", "in", valid_no_mix.ids)]
+            loc_ordered_by_qty = [
+                item["location_id"][0]
+                for item in StockQuant.read_group(
+                    domain_quant,
+                    ["location_id", "quantity"],
+                    ["location_id"],
+                    orderby="quantity",
+                )
+                if (float_compare(item["quantity"], 0, precision_digits=2) > 0)
+            ]
+        valid_location_ids = set(valid_locations.ids) - set(loc_ordered_by_qty)
+        valid_locations = self.browse(loc_ordered_by_qty)
```

Comment:
instead of intersection can't you browse once like this?
```py
ids = loc_ordered_by_qty + set([id_ for id_ in self.ids if id_ in valid_location_ids])
self.browse(ids)
```
```

Reply:
`set`will loose the initial order....